From 2c894a76c7fb002ed276d1d82552fb4e91bb3a78 Mon Sep 17 00:00:00 2001
From: Carlos Garnacho <carlos@lanedo.com>
Date: Fri, 10 May 2013 18:29:05 +0200
Subject: [PATCH 60/68] iconfactory: Add gtk_cairo_set_source_icon_set()

As GtkIconSet remains a good candidate to delay pixbuf lookup until
the moment of rendering, a cairo helper function has been added
to easy enabling support for high resolution icons.
---
 gtk/gtkiconfactory.c |   32 ++++++++++++++++++++++++++++++++
 gtk/gtkiconfactory.h |    9 +++++++++
 2 files changed, 41 insertions(+)

diff --git a/gtk/gtkiconfactory.c b/gtk/gtkiconfactory.c
index 41d1ee7..ab9a212 100644
--- a/gtk/gtkiconfactory.c
+++ b/gtk/gtkiconfactory.c
@@ -3127,5 +3127,37 @@ gtk_icon_source_get_filename (const GtkIconSource *source)

 #endif

+void
+gtk_cairo_set_source_icon_set (cairo_t         *cr,
+                               GtkWidget       *widget,
+                               GtkIconSet      *icon_set,
+                               GtkIconSize      size,
+                               gdouble          scale,
+                               gdouble          icon_x,
+                               gdouble          icon_y)
+{
+  cairo_pattern_t *pattern;
+  cairo_matrix_t matrix;
+  GdkPixbuf *pixbuf;
+
+  g_return_if_fail (GTK_IS_WIDGET (widget));
+  g_return_if_fail (cr != NULL);
+  g_return_if_fail (icon_set != NULL);
+
+  pixbuf = gtk_icon_set_render_icon_scaled (icon_set,
+                                            gtk_widget_get_style (widget),
+                                            gtk_widget_get_direction (widget),
+                                            gtk_widget_get_state (widget),
+                                            size, widget, NULL, &scale);
+  gdk_cairo_set_source_pixbuf (cr, pixbuf, 0, 0);
+
+  pattern = cairo_get_source (cr);
+  cairo_matrix_init_scale (&matrix, scale, scale);
+  cairo_matrix_translate (&matrix, -icon_x, -icon_y);
+  cairo_pattern_set_matrix (pattern, &matrix);
+
+  g_object_unref (pixbuf);
+}
+
 #define __GTK_ICON_FACTORY_C__
 #include "gtkaliasdef.c"
diff --git a/gtk/gtkiconfactory.h b/gtk/gtkiconfactory.h
index d646ed9..67e318a 100644
--- a/gtk/gtkiconfactory.h
+++ b/gtk/gtkiconfactory.h
@@ -196,6 +196,15 @@ GtkStateType     gtk_icon_source_get_state                (const GtkIconSource *
 GtkIconSize      gtk_icon_source_get_size                 (const GtkIconSource *source);
 gdouble          gtk_icon_source_get_scale                (const GtkIconSource *source);

+/* Cairo helper */
+void             gtk_cairo_set_source_icon_set            (cairo_t             *cr,
+                                                           GtkWidget           *widget,
+                                                           GtkIconSet          *icon_set,
+                                                           GtkIconSize          size,
+                                                           gdouble              scale,
+                                                           gdouble              icon_x,
+                                                           gdouble              icon_y);
+
 /* ignore this */
 void _gtk_icon_set_invalidate_caches (void);
 GList* _gtk_icon_factory_list_ids (void);
--
1.7.10.2 (Apple Git-33)
